src/replication.c

```diff
@@ -126,37 +126,215 @@ static int handle_mgmt_event_fd(replica_t *replica);
 }
 
 /*
- * Update_volstate enables spec and initialize io_seq number
- * if (n_replica >= c_factor)  or
- *    (n_replica >= MAX (c_factor, r_factor - c_factor + 1).
- *  n_replica = number of replica connected
- *  c_factor = consistency factor
- *  r_factor = replication_factor
+ * Case 1: Rebuild from healthy replica
+ * Case 2: Mesh rebuild i.e all replicas are downgrade
+ */
+static int 
+prepare_for_rebuild(spec_t *spec, replica_t *target_replica,
+    replica_t *healthy_replica)
+{
+	int ret = 0;
+	uint64_t num = 1;
+	replica_t *replica;
+	zvol_io_hdr_t *rmgmtio = NULL;
+	uint64_t data_len = strlen(spec->volname) + 1;
+	zvol_op_code_t mgmt_opcode = ZVOL_OPCODE_PREPARE_FOR_REBUILD;
+	mgmt_cmd_t *mgmt_cmd;
+
+	// Case 1
+	if (healthy_replica != NULL) {
+
+		mgmt_cmd = malloc(sizeof (mgmt_cmd_t));
+		BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
+
+		mgmt_cmd->io_hdr = rmgmtio;
+		mgmt_cmd->io_bytes = 0;
+		mgmt_cmd->data = (char *)malloc(data_len);
+		snprintf((char *)mgmt_cmd->data, data_len, "%s", spec->volname);
+		mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+		spec->rebuild_replica_count++;  
+	
+		MTX_LOCK(&healthy_replica->r_mtx);
+		TAILQ_INSERT_TAIL(&healthy_replica->mgmt_cmd_queue, mgmt_cmd,
+		    mgmt_cmd_next);
+		MTX_UNLOCK(&healthy_replica->r_mtx);
+
+		if (write(healthy_replica->mgmt_eventfd1, &num,
+		    sizeof (num)) != sizeof (num)) {
+			REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+			    "for replica(%p)\n", replica);
+			ret = -1;
+		}
+		return ret;
+	}
+
+	// Case 2
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica == target_replica)
+			continue;
+
+		spec->rebuild_replica_count++;  
```

Comment:
similar to rcmd_t (which is about data command sent to replica) and rcomm_cmd_t (which is common structure contains details about all rcmd_t's sent to replicas), mgmt_cmd (mgmt command sent to replica) have rcommon_mgmt_cmd_t. This can be helpful here, to replace newly added rebuild related variables of spec.